### DISCUSSION TITLE
emu.register_menu documentation?

### DISCUSSION THREAD

**Discussion by Jakobud** [mamedev/discussions#159] [Q&A] [answered]
Is there any documentation around `emu.register_menu` and the populate and callback functions that go with them? I'm looking through the source and plugin examples and can't figure it out 100%. For example:

timer plugin:
```
local function menu_populate()
		lastupdate = os.time()
		local refname = (reference == 0) and _p('plugin-timer', 'Wall clock') or _p('plugin-timer', 'Emulated time')
		local time = (reference == 0) and (lastupdate - start_time) or manager.machine.time.seconds
		local total = (reference == 0) and (total_time + time) or (manager.machine.time + emu_total).seconds
		return
			{
				{ _p("plugin-timer", "Reference"), refname, (reference == 0) and 'r' or 'l' },
				{ '---', '', '' },
				{ _p("plugin-timer", "Current time"), sectohms(time), "off" },
				{ _p("plugin-timer", "Total time"), sectohms(total), "off" },
				{ _p("plugin-timer", "Play Count"), tostring(play_count), "off" } },
			highlight,
			"idle"
	end
```
The function returns a table, and then a local variable called `highlight` that is nil I think, and then the string "idle". Is there any documentation around any of this? Also `{ _p("plugin-timer", "Reference"), refname, (reference == 0) and 'r' or 'l' },` what is that? Is that a menu option that you define a string name for it, the variable that it stores and the default value?

timecode plugin
```
It has `table.insert(result, { _p('plugin-timecode', 'Frame numbers'), frame_mode_val, (frame_mode > 0) and 'l' or 'r' })` there is that `l` or `r` thing again. Is it setting the value (default value?) to `l` if `frame_mode > 0` otherwise `l` ?

Side note, is `_p('plugin-name', 'some-string')` some sort of localization thing? Any documentation on this?

**Comment by cuavas** [comment, marked as answer]
It's poorly designed and needs to be replaced. It won't be documented in its current form because it's basically unsupportable. You can look at the source of the menu class behind it, or follow examples in existing plugins.

The three return values are:
* Table of menu items
* Index of item to move highlight to
* Menu processing flags

Each entry in the table of items is:
* Item text (some values are magic, like `---` for producing a separator)
* Value text
* Item flags (e.g. `off` for a disabled item or `l` and `r` to show the arrows for adjustable values)
